File: MiniDOMFinanziaria/Finanziaria/Prodotti/TabelleSpeseXProdotto/CProdottoXTabellaSpesa.cs
using System;
using DMD;
using DMD.XML;
using minidom;
using static minidom.Sistema;


namespace minidom
{
    public partial class Finanziaria
    {
        /// <summary>
    /// Relazione Prodotti x Tabellea Spese ( 1 - molti )
    /// </summary>
    /// <remarks></remarks>
        [Serializable]
        public class CProdottoXTabellaSpesa : Databases.DBObject, IComparable, ICloneable
        {
            private string m_Nome; // Descrizione
            private int m_IDProdotto; // ID del prodotto associato
            [NonSerialized]
            private CCQSPDProdotto m_Prodotto; // Oggetto Prodotto Associato
            private int m_IDTabellaSpese; // ID della tablla spese
            [NonSerialized]
            private CTabellaSpese m_TabellaSpese;
            private int m_Flags;
            private DateTime? m_DataInizio;
            private DateTime? m_DataFine;

            public CProdottoXTabellaSpesa()
            {
                m_Nome = "";
                m_IDProdotto = 0;
                m_Prodotto = null;
                m_IDTabellaSpese = 0;
                m_TabellaSpese = null;
                m_Flags = 0;
                m_DataInizio = default;
                m_DataFine = default;
            }

            public override CModulesClass GetModule()
            {
                return null;
            }

            public string Nome
            {
                get
                {
                    return m_Nome;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_Nome;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Nome = value;
                    DoChanged("Nome", value, oldValue);
                }
            }

            /// <summary>
        /// ID del prodotto associato
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
            public int IDProdotto
            {
                get
                {
                    return DBUtils.GetID(m_Prodotto, m_IDProdotto);
                }

                set
                {
                    int oldValue = IDProdotto;
                    if (oldValue == value)
                        return;
                    m_IDProdotto = value;
                    m_Prodotto = null;
                    DoChanged("IDProdotto", value, oldValue);
                }
            }

            /// <summary>
        /// Oggetto Prodotto Associato
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
            public CCQSPDProdotto Prodotto
            {
                get
                {
                    if (m_Prodotto is null)
                        m_Prodotto = Prodotti.GetItemById(m_IDProdotto);
                    return m_Prodotto;
                }

                set
                {
                    var oldValue = Prodotto;
                    if (oldValue == value)
                        return;
                    m_Prodotto = value;
                    m_IDProdotto = DBUtils.GetID(value);
                    DoChanged("Prodotto", value, oldValue);
                }
            }

            protected internal virtual void SetProdotto(CCQSPDProdotto value)
            {
                m_Prodotto = value;
                m_IDProdotto = DBUtils.GetID(value);
            }

            public int IDTabellaSpese
            {
                get
                {
                    return DBUtils.GetID(m_TabellaSpese, m_IDTabellaSpese);
                }

                set
                {
                    int oldValue = IDTabellaSpese;
                    if (oldValue == value)
                        return;
                    m_IDTabellaSpese = value;
                    m_TabellaSpese = null;
                    DoChanged("IDTabellaSpese", value, oldValue);
                }
            }

            public CTabellaSpese TabellaSpese
            {
                get
                {
                    if (m_TabellaSpese is null)
                        m_TabellaSpese = TabelleSpese.GetItemById(m_IDTabellaSpese);
                    return m_TabellaSpese;
                }

                set
                {
                    var oldValue = TabellaSpese;
                    if (oldValue == value)
                        return;
                    m_TabellaSpese = value;
                    m_IDTabellaSpese = DBUtils.GetID(value);
                    DoChanged("TabellaSpese", value, oldValue);
                }
            }

            public int Flags
            {
                get
                {
                    return m_Flags;
                }

                set
                {
                    int oldValue = m_Flags;
                    if (oldValue == value)
                        return;
                    m_Flags = value;
                    DoChanged("Flags", value, oldValue);
                }
            }

            public DateTime? DataInizio
            {
                get
                {
                    return m_DataInizio;
                }

                set
                {
                    var oldValue = m_DataInizio;
                    if (DMD.DateUtils.Compare(value, oldValue) == 0)
                        return;
                    m_DataInizio = value;
                    DoChanged("DataInizio", value, oldValue);
                }
            }

            public DateTime? DataFine
            {
                get
                {
                    return m_DataFine;
                }

                set
                {
                    var oldValue = m_DataFine;
                    if (DMD.DateUtils.Compare(value, oldValue) == 0)
                        return;
                    m_DataFine = value;
                    DoChanged("DataFine", value, oldValue);
                }
            }

            public override string GetTableName()
            {
                return "tbl_FIN_ProdXTabSpes";
            }

            protected override void OnAfterSave(SystemEvent e)
            {
                base.OnAfterSave(e);
                foreach (CCQSPDProdotto p in Prodotti.LoadAll())
                    p.InvalidateTabelleSpese();
            }

            public bool IsValid()
            {
                return IsValid(DMD.DateUtils.Now());
            }

            public bool IsValid(DateTime at)
            {
                return DMD.DateUtils.CheckBetween(at, m_DataInizio, m_DataFine);
            }

            protected override Databases.CDBConnection GetConnection()
            {
                return Database;
            }

            protected override bool LoadFromRecordset(DBReader reader)
            {
                m_Nome = reader.Read("Nome", this.m_Nome);
                m_IDProdotto = reader.Read("IDProdotto", this.m_IDProdotto);
                m_IDTabellaSpese = reader.Read("IDTabellaSpese", this.m_IDTabellaSpese);
                m_Flags = reader.Read("Flags", this.m_Flags);
                m_DataInizio = reader.Read("DataInizio", this.m_DataInizio);
                m_DataFine = reader.Read("DataFine", this.m_DataFine);
                return base.LoadFromRecordset(reader);
            }

            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("Nome", m_Nome);
                writer.Write("IDProdotto", IDProdotto);
                writer.Write("IDTabellaSpese", IDTabellaSpese);
                writer.Write("Flags", m_Flags);
                writer.Write("DataInizio", m_DataInizio);
                writer.Write("DataFine", m_DataFine);
                return base.SaveToRecordset(writer);
            }

            public override string ToString()
            {
                return m_Nome;
            }

            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Nome", m_Nome);
                writer.WriteAttribute("IDProdotto", IDProdotto);
                writer.WriteAttribute("IDTabellaSpese", IDTabellaSpese);
                writer.WriteAttribute("Flags", m_Flags);
                writer.WriteAttribute("DataInizio", m_DataInizio);
                writer.WriteAttribute("DataFine", m_DataFine);
                base.XMLSerialize(writer);
            }

            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Nome":
                        {
                            m_Nome = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "IDProdotto":
                        {
                            m_IDProdotto = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "IDTabellaSpese":
                        {
                            m_IDTabellaSpese = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Flags":
                        {
                            m_Flags = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "DataInizio":
                        {
                            m_DataInizio = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "DataFine":
                        {
                            m_DataFine = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

            public int CompareTo(CProdottoXTabellaSpesa other)
            {
                return DMD.Strings.Compare(m_Nome, other.m_Nome, true);
            }

            public int CompareTo(object other)
            {
                return CompareTo((CProdottoXTabellaSpesa)other);
            }

            public int _CompareTo(object other) => CompareTo(other);

            public object Clone()
            {
                return MemberwiseClone();
            }
        }
    }
}

File: MiniDOMFinanziaria/Finanziaria/Prodotti/TabelleSpeseXProdotto/CProdottoXTabellaSpesa.vb
Imports minidom.Databases
Imports minidom.Sistema

Partial Public Class Finanziaria
    ''' <summary>
    ''' Relazione Prodotti x Tabellea Spese ( 1 - molti )
    ''' </summary>
    ''' <remarks></remarks>
    <Serializable>
    Public Class CProdottoXTabellaSpesa
        Inherits DBObject
        Implements IComparable, ICloneable

        Private m_Nome As String 'Descrizione
        Private m_IDProdotto As Integer 'ID del prodotto associato
        <NonSerialized> Private m_Prodotto As CCQSPDProdotto 'Oggetto Prodotto Associato
        Private m_IDTabellaSpese As Integer 'ID della tablla spese
        <NonSerialized> Private m_TabellaSpese As CTabellaSpese
        Private m_Flags As Integer
        Private m_DataInizio As DateTime?
        Private m_DataFine As DateTime?

        Public Sub New()
            Me.m_Nome = ""
            Me.m_IDProdotto = 0
            Me.m_Prodotto = Nothing
            Me.m_IDTabellaSpese = 0
            Me.m_TabellaSpese = Nothing
            Me.m_Flags = 0
            Me.m_DataInizio = Nothing
            Me.m_DataFine = Nothing
        End Sub

        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Public Property Nome As String
            Get
                Return Me.m_Nome
            End Get
            Set(value As String)
                value = Strings.Trim(value)
                Dim oldValue As String = Me.m_Nome
                If (oldValue = value) Then Exit Property
                Me.m_Nome = value
                Me.DoChanged("Nome", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' ID del prodotto associato
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property IDProdotto As Integer
            Get
                Return GetID(Me.m_Prodotto, Me.m_IDProdotto)
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.IDProdotto
                If (oldValue = value) Then Exit Property
                Me.m_IDProdotto = value
                Me.m_Prodotto = Nothing
                Me.DoChanged("IDProdotto", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Oggetto Prodotto Associato
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Prodotto As CCQSPDProdotto
            Get
                If (Me.m_Prodotto Is Nothing) Then Me.m_Prodotto = Finanziaria.Prodotti.GetItemById(Me.m_IDProdotto)
                Return Me.m_Prodotto
            End Get
            Set(value As CCQSPDProdotto)
                Dim oldValue As CCQSPDProdotto = Me.Prodotto
                If (oldValue = value) Then Exit Property
                Me.m_Prodotto = value
                Me.m_IDProdotto = GetID(value)
                Me.DoChanged("Prodotto", value, oldValue)
            End Set
        End Property


        Protected Friend Overridable Sub SetProdotto(ByVal value As CCQSPDProdotto)
            Me.m_Prodotto = value
            Me.m_IDProdotto = GetID(value)
        End Sub

        Public Property IDTabellaSpese As Integer
            Get
                Return GetID(Me.m_TabellaSpese, Me.m_IDTabellaSpese)
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.IDTabellaSpese
                If (oldValue = value) Then Exit Property
                Me.m_IDTabellaSpese = value
                Me.m_TabellaSpese = Nothing
                Me.DoChanged("IDTabellaSpese", value, oldValue)
            End Set
        End Property

        Public Property TabellaSpese As CTabellaSpese
            Get
                If (Me.m_TabellaSpese Is Nothing) Then Me.m_TabellaSpese = Finanziaria.TabelleSpese.GetItemById(Me.m_IDTabellaSpese)
                Return Me.m_TabellaSpese
            End Get
            Set(value As CTabellaSpese)
                Dim oldValue As CTabellaSpese = Me.TabellaSpese
                If (oldValue = value) Then Exit Property
                Me.m_TabellaSpese = value
                Me.m_IDTabellaSpese = GetID(value)
                Me.DoChanged("TabellaSpese", value, oldValue)
            End Set
        End Property

        Public Property Flags As Integer
            Get
                Return Me.m_Flags
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.m_Flags
                If (oldValue = value) Then Return
                Me.m_Flags = value
                Me.DoChanged("Flags", value, oldValue)
            End Set
        End Property

        Public Property DataInizio As DateTime?
            Get
                Return Me.m_DataInizio
            End Get
            Set(value As DateTime?)
                Dim oldValue As DateTime? = Me.m_DataInizio
                If (DateUtils.Compare(value, oldValue) = 0) Then Return
                Me.m_DataInizio = value
                Me.DoChanged("DataInizio", value, oldValue)
            End Set
        End Property

        Public Property DataFine As DateTime?
            Get
                Return Me.m_DataFine
            End Get
            Set(value As DateTime?)
                Dim oldValue As DateTime? = Me.m_DataFine
                If (DateUtils.Compare(value, oldValue) = 0) Then Return
                Me.m_DataFine = value
                Me.DoChanged("DataFine", value, oldValue)
            End Set
        End Property

        Public Overrides Function GetTableName() As String
            Return "tbl_FIN_ProdXTabSpes"
        End Function

        Protected Overrides Sub OnAfterSave(e As SystemEvent)
            MyBase.OnAfterSave(e)
            For Each p As CCQSPDProdotto In Finanziaria.Prodotti.LoadAll
                p.InvalidateTabelleSpese()
            Next
        End Sub

        Public Function IsValid() As Boolean
            Return Me.IsValid(DateUtils.Now)
        End Function

        Public Function IsValid(ByVal at As DateTime) As Boolean
            Return DateUtils.CheckBetween(at, Me.m_DataInizio, Me.m_DataFine)
        End Function


        Protected Overrides Function GetConnection() As CDBConnection
            Return Finanziaria.Database
        End Function

        Protected Overrides Function LoadFromRecordset(ByVal reader As DBReader) As Boolean
            Me.m_Nome = reader.Read("Nome", Me.m_Nome)
            Me.m_IDProdotto = reader.Read("IDProdotto", Me.m_IDProdotto)
            Me.m_IDTabellaSpese = reader.Read("IDTabellaSpese", Me.m_IDTabellaSpese)
            Me.m_Flags = reader.Read("Flags", Me.m_Flags)
            Me.m_DataInizio = reader.Read("DataInizio", Me.m_DataInizio)
            Me.m_DataFine = reader.Read("DataFine", Me.m_DataFine)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(ByVal writer As DBWriter) As Boolean
            writer.Write("Nome", Me.m_Nome)
            writer.Write("IDProdotto", Me.IDProdotto)
            writer.Write("IDTabellaSpese", Me.IDTabellaSpese)
            writer.Write("Flags", Me.m_Flags)
            writer.Write("DataInizio", Me.m_DataInizio)
            writer.Write("DataFine", Me.m_DataFine)
            Return MyBase.SaveToRecordset(writer)
        End Function

        Public Overrides Function ToString() As String
            Return Me.m_Nome
        End Function

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("Nome", Me.m_Nome)
            writer.WriteAttribute("IDProdotto", Me.IDProdotto)
            writer.WriteAttribute("IDTabellaSpese", Me.IDTabellaSpese)
            writer.WriteAttribute("Flags", Me.m_Flags)
            writer.WriteAttribute("DataInizio", Me.m_DataInizio)
            writer.WriteAttribute("DataFine", Me.m_DataFine)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "Nome" : Me.m_Nome = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "IDProdotto" : Me.m_IDProdotto = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "IDTabellaSpese" : Me.m_IDTabellaSpese = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Flags" : Me.m_Flags = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "DataInizio" : Me.m_DataInizio = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "DataFine" : Me.m_DataFine = XML.Utils.Serializer.DeserializeDate(fieldValue)

                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub


        Public Function CompareTo(other As CProdottoXTabellaSpesa) As Integer
            Return Strings.Compare(Me.m_Nome, other.m_Nome, CompareMethod.Text)
        End Function

        Public Function _CompareTo(other As Object) As Integer Implements IComparable.CompareTo
            Return Me.CompareTo(CType(other, CProdottoXTabellaSpesa))
        End Function

        Public Function Clone() As Object Implements ICloneable.Clone
            Return Me.MemberwiseClone
        End Function
    End Class


End Class